File: GenomicPrediction/interfaces/python/sample_usage.py
# interfaces/python/sample_usage.py
# -------------------------------------
#
# This script demonstrates how to call the GenomicPrediction.jl package from Python
# using the PyJulia library.

import julia
from julia import Pkg

def main():
    print("Initializing Julia runtime...")
    julia.Julia(compiled_modules=False)

    print("Activating Julia environment for GenomicPrediction.jl...")
    Pkg.activate("../../")

    print("Loading GenomicPrediction.jl package...")
    from julia import GenomicPrediction as GP

    # --- 1. Load Data ---
    print("\n--- Step 1: Loading Data ---")
    geno_path = "../sample_data/genotypes.csv"
    pheno_path = "../sample_data/phenotypes.csv"

    # Create dummy data for the example
    from julia import CSV, DataFrame
    CSV.write(geno_path, DataFrame(ID=range(1, 6), m1=[0,1,2,0,1], m2=[2,1,0,2,1]))
    CSV.write(pheno_path, DataFrame(ID=range(1, 6), y=[1.1, 1.9, 3.2, 1.2, 2.3]))

    mock_data = GP.load_csv(geno_path, pheno_path, header_geno=True, header_pheno=True)
    print("Data loaded successfully.")

    # --- 2. Initialize and Train a Model ---
    print("\n--- Step 2: Training GBLUP Model ---")
    model = GP.GBLUPModel(10.0)
    GP.fit_b(model, mock_data) # fit! is aliased to fit_b
    print("Model training complete.")

    # --- 3. Save and Load Model ---
    print("\n--- Step 3: Saving and Reloading Model ---")
    temp_path = "temp_model.bson"
    GP.save_model(model, temp_path)
    print(f"Model saved to {temp_path}")
    loaded_model = GP.load_model(temp_path)
    print("Model successfully reloaded.")

    # --- 4. Make Predictions ---
    print("\n--- Step 4: Making Predictions ---")
    genotypes_df = mock_data.genotypes
    predictions = GP.predict(loaded_model, genotypes_df)

    print("Predictions obtained.")

    # --- 5. Display Results ---
    print("\n--- Step 5: Results ---")
    predictions_list = list(predictions)
    print(f"Predicted values: {predictions_list}")

if __name__ == "__main__":
    main()

File: GenomicPrediction/interfaces/R/sample_usage.R
# interfaces/R/sample_usage.R
# -----------------------------
#
# This script demonstrates how to call the GenomicPrediction.jl package from R
# using the JuliaCall library.

# --- 1. Setup ---
# First, ensure you have JuliaCall installed in R:
# install.packages("JuliaCall")

library(JuliaCall)

# --- 2. Initialize Julia ---
cat("Initializing Julia runtime...\n")
# Point to your Julia installation if not in PATH
# julia_setup(JULIA_HOME = "/path/to/julia/bin")
julia_setup()

# --- 3. Activate Project and Load Package ---
cat("Activating Julia project and loading GenomicPrediction.jl...\n")
# Go up two directories to the root of the GenomicPrediction.jl package
julia_command("using Pkg; Pkg.activate('../../')")

# Load the module. We can assign it to an R object for convenience.
julia_library("GenomicPrediction")

# --- 4. Prepare Data ---
cat("\n--- Step 4: Preparing Data ---\n")
# Create dummy data files for the example
geno_path <- "../sample_data/genotypes.csv"
pheno_path <- "../sample_data/phenotypes.csv"
write.csv(data.frame(ID = 1:5, m1 = c(0,1,2,0,1), m2 = c(2,1,0,2,1)), geno_path, row.names = FALSE)
write.csv(data.frame(ID = 1:5, y = c(1.1, 1.9, 3.2, 1.2, 2.3)), pheno_path, row.names = FALSE)

# Call the Julia function to load data
# The 'julia_call' function executes a Julia function with R arguments
mock_data <- julia_call("GenomicPrediction.load_csv", geno_path, pheno_path, header_geno=TRUE, header_pheno=TRUE)
cat("Data loaded successfully into a Julia object.\n")


# --- 5. Initialize and Train Model ---
cat("\n--- Step 5: Training GBLUP Model ---\n")
# Create a GBLUP model instance
gblup_model <- julia_call("GenomicPrediction.GBLUPModel", 10.0)

# Train the model
# Note: Julia functions with '!' are mutated. In R, we just call them.
julia_call("GenomicPrediction.fit!", gblup_model, mock_data)
cat("Model training complete.\n")


# --- 6. Save and Load Model ---
cat("\n--- Step 6: Saving and Reloading Model ---\n")
temp_path <- "temp_model.bson"
julia_call("GenomicPrediction.save_model", gblup_model, temp_path)
cat(paste("Model saved to", temp_path, "\n"))

loaded_model <- julia_call("GenomicPrediction.load_model", temp_path)
cat("Model successfully reloaded.\n")

# --- 7. Make Predictions ---
cat("\n--- Step 7: Making Predictions ---\n")
# Extract the genotype DataFrame from the Julia object
genotypes_df <- mock_data$genotypes

# Predict using the reloaded model
predictions_jl <- julia_call("GenomicPrediction.predict", loaded_model, genotypes_df)

# Convert Julia vector to R vector
predictions_r <- julia_eval("x -> collect(x)", predictions_jl)

cat("Predictions obtained and converted to R vector.\n")

# --- 8. Display Results ---
cat("\n--- Step 8: Results ---\n")
cat("Predicted values:\n")
print(predictions_r)

# Clean up temporary files
file.remove(geno_path)
file.remove(pheno_path)
file.remove(temp_path)
